subsys/usb/device/class/audio/audio.c

```diff
@@ -541,6 +541,73 @@ static int handle_fu_mute_req(struct usb_audio_dev_data *audio_dev_data,
 	return -EINVAL;
 }
 
+/**
+ * @brief Handler for feature unit volume control requests.
+ *
+ * This function handles feature unit volume control request.
+ *
+ * @param audio_dev_data USB audio device data.
+ * @param setup          Information about the executed request.
+ * @param len		 Size of the buffer.
+ * @param data		 Buffer containing the request result.
+ * @param evt		 Feature Unit Event info.
+ * @param device	 Device part that has been addressed. Applicable for
+ *			 bidirectional device.
+ *
+ * @return 0 if successful, negative errno otherwise.
+ */
+
+static int handle_fu_volume_req(struct usb_audio_dev_data *audio_dev_data,
+			      struct usb_setup_packet *setup,
+			      int32_t *len, uint8_t **data,
+			      struct usb_audio_fu_evt *evt,
+			      uint8_t device)
+{
+	uint8_t ch = (setup->wValue) & 0xFF;
+	uint8_t ch_cnt = audio_dev_data->ch_cnt[device];
+	uint8_t *controls = audio_dev_data->controls[device];
+	uint8_t *control_val = &controls[POS(VOLUME, ch, ch_cnt)];
+	uint16_t temp_vol = 0;
+	uint16_t temp_res = 0;
+
+	if (usb_reqtype_is_to_device(setup)) {
+		/* Check if *len has valid value */
+		if (*len != LEN(1, VOLUME)) {
+			LOG_ERR("*len: %d, LEN(1, VOLUME): %d", *len, LEN(1, VOLUME));
+			return -EINVAL;
+		}
+		if (setup->bRequest == USB_AUDIO_SET_CUR) {
+			evt->val = control_val;
+			evt->val_len = *len;
+			memcpy(control_val, *data, *len);
+			return 0;
+		}
```

Comment:
The value is not checked against the min and max. Additionally it does not honor the resolution rounding.

Values in USB are little-endian, not target-device-endian. You are missing necessary conversions here (which evalue to no-op on little-endian targets).

Reply:
Thanks a lot.  I fixed it.